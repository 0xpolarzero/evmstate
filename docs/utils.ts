import { isHex } from "tevm";
import clsx, { type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const maxBytesLength = 64;
export const stringify = (value: any, space: number = 2) => {
  return JSON.stringify(
    value,
    (_, v) => {
      if (typeof v === "bigint") return `__bigint__${v.toString()}`;
      if (isHex(v) && (v.length - 2) / 2 > maxBytesLength)
        return `__long_hex__${v.slice(0, maxBytesLength + 2)}...${v.slice(-maxBytesLength)}`;
      return v;
    },
    space,
  );
};
