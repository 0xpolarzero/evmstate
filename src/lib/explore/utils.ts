import { Hex, toHex } from "tevm";
import { ByteArray, ToHexParameters } from "viem";

export const max = <T extends bigint | number>(...nums: Array<T>): T =>
  // @ts-expect-error bigint/number
  nums.reduce((max, num) => (num > max ? num : max), -Infinity);

/**
 * Converts a value to a hex string and ensures it has an even number of characters (full bytes)
 *
 * @param value The value to convert to hex
 * @param options Optional configuration options passed to tevm's toHex
 * @returns A hex string with full bytes
 */
export const toHexFullBytes = (value: string | number | bigint | boolean | ByteArray, opts?: ToHexParameters): Hex => {
  // Use tevm's toHex function
  const hex = toHex(value, opts);

  // If the hex string has an odd number of characters after the 0x prefix,
  // pad it with a leading zero to make it even
  if ((hex.length - 2) % 2 !== 0) return ("0x0" + hex.slice(2)) as Hex;
  return hex;
};
