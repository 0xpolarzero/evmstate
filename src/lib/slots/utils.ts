import { Address, decodeAbiParameters, encodeAbiParameters, Hex, isHex, toHex } from "tevm";
import { abi } from "@shazow/whatsabi";
import { AbiType, AbiTypeToPrimitiveType } from "abitype";
import { padHex } from "viem";

import { MappingKey } from "@/lib/slots/types";

/* -------------------------------------------------------------------------- */
/*                                  MAPPINGS                                  */
/* -------------------------------------------------------------------------- */

/** Extract values from a transaction trace that might be used as mapping keys */
// TODO: refactor to navigate the stack trace and identify both potential keys and array indexes
/**
 * Typical access patterns:
 *
 * Mapping
 *
 * - KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key, uint256(mappingSlot)))
 * - -> key is the input to the hash so we might even be able to retrieve the slot and compare directly
 * - KECCAK256 → KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key2, keccak256(abi.encode(key1, uint256(mappingSlot)))))
 * - -> same here we get each key from each keccak input
 *
 * Array
 *
 * - KECCAK256 → ADD → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(uint256(arraySlot)) + index
 *
 * As reference:
 *
 * - Struct in array: KECCAK256 → ADD → ADD → SLOAD/SSTORE Slot calculation: (keccak256(uint256(arraySlot)) + index) +
 *   memberOffset
 * - Mapping to Struct with Mapping: KECCAK256 → ADD → KECCAK256 → SLOAD/SSTORE Slot calculation:
 *   keccak256(abi.encode(key2, (keccak256(abi.encode(key1, uint256(mappingSlot))) + memberOffset)))
 */
export const extractPotentialKeys = (
  trace: {
    uniqueStackValues?: Array<string>;
    relevantOps?: Array<{
      op: string;
      stack: Array<string>;
    }>;
  },
  addresses: Array<Address>,
  abiFunctions: Array<abi.ABIFunction>,
  txData?: Hex,
): MappingKey[] => {
  const keys: MappingKey[] = [];

  // Add touched addresses
  addresses.forEach((address) => {
    keys.push({
      hex: padHex(address, { size: 32 }),
      decoded: address,
      type: "address",
    });
  });

  // Extract parameters from transaction data
  if (txData && txData.length > 10) {
    const selector = txData.slice(0, 10);
    const inputs = abiFunctions.find((fn) => fn.selector === selector)?.inputs;

    if (inputs) {
      // Decode function inputs
      const params = decodeAbiParameters(inputs, `0x${txData.slice(10)}`);

      params.forEach((param, index) => {
        // If it's an array, add each element as a key
        if (Array.isArray(param)) {
          param.forEach((p) => {
            const type = inputs[index].type.replace("[]", "") as AbiType;

            if (type && type !== "string" && type !== "bytes") {
              keys.push({
                hex: padHex(encodeAbiParameters([{ type }], [p]), { size: 32 }),
                decoded: p as AbiTypeToPrimitiveType<typeof type>,
                type,
              });
            }
          });
        } else {
          // Otherwise just add the key straight up
          const type = inputs[index].type as AbiType;

          if (type && type !== "string" && type !== "bytes") {
            keys.push({
              hex: padHex(encodeAbiParameters([{ type }], [param]), { size: 32 }),
              decoded: param as AbiTypeToPrimitiveType<typeof type>,
              type,
            });
          }
        }
      });
    }
  }

  // Process stack values from the trace
  if (trace.uniqueStackValues?.length) {
    // Process unique stack values directly
    for (const stackValue of trace.uniqueStackValues) {
      keys.push({
        hex: isHex(stackValue) ? padHex(stackValue, { size: 32 }) : toHex(stackValue, { size: 32 }),
        type: undefined,
      });
    }
  }

  // Deduplicate keys
  const uniqueMap = new Map();
  // Add the new key only if it's not already in the map (and don't replace a key with a defined type)
  keys.forEach((k) => {
    if (!uniqueMap.has(k.hex) || k.type) uniqueMap.set(k.hex, k);
  });

  return Array.from(uniqueMap.values()).sort((a, b) => {
    // prefer address as it's more likely to be a key
    if (a.type === "address") return -1;
    if (b.type === "address") return 1;
    // prefer defined types
    if (a.type === undefined) return 1;
    if (b.type === undefined) return -1;

    return 0;
  });
};

/* -------------------------------------------------------------------------- */
/*                                    UTILS                                   */
/* -------------------------------------------------------------------------- */

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
