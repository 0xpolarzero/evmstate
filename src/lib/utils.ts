import { Address } from "tevm";

export const uniqueAddresses = (addresses: Array<Address | undefined>): Array<Address> => {
  let existingAddresses = new Set<string>();

  return addresses.filter((address) => {
    if (!address || existingAddresses.has(address.toLowerCase())) return false;
    existingAddresses.add(address.toLowerCase());
    return true;
  }) as Address[];
};
