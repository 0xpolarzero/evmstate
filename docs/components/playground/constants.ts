import { EthjsAddress } from "tevm/utils";
import type { ExtractAbiFunctions } from "abitype";

import * as LAYOUTS from "@test/codegen/layouts/index.js";
import * as CONTRACTS from "@test/contracts/index.js";

export const contract = CONTRACTS["Playground"].withAddress(
  EthjsAddress.fromString("0x987C2AF139EAEaBdF8D6d3d1723C1883bEa1f2AF").toString(),
);
export const layout = LAYOUTS["Playground"];
export const callerAddress = EthjsAddress.fromString("0xCa11e40000000000000000000000000000000000");
export const localStorageKey = "EVMSTATE_PLAYGROUND_STATE";

export const functionDescriptions: Record<ExtractAbiFunctions<typeof contract.abi, "nonpayable">["name"], string> = {
  addValue: "Adds to a dynamic array",
  addUser: "Adds a user struct to a mapping and array",
  toggleUserActive: "Toggles a boolean in a struct within a mapping",
  setBalance: "Updates a simple mapping value",
  setAllowance: "Updates a nested mapping value",
  addTransaction: "Adds to a dynamic array within a mapping",
  updateBasicValues: "Updates primitive types (uint256, bool)",
  updatePackedValues: "Updates packed storage variables (uint8, uint16, uint32, bool)",
  setStringAndBytes: "Updates string and bytes storage",
  setFixedValue: "Updates a fixed-size array element",
};
