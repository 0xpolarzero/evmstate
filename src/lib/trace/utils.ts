import {
  Abi,
  BlockTag,
  ContractFunctionName,
  createMemoryClient,
  encodeFunctionData,
  Hex,
  http,
  MemoryClient,
} from "tevm";
import { Common } from "tevm/common";

import { debug } from "@/debug";
import { ExploreStorageConfig } from "@/lib/explore/config";
import { TraceStorageAccessOptions, TraceStorageAccessTxParams, TraceStorageAccessTxWithData } from "@/lib/trace/types";

/** Creates a Tevm client from the provided options */
export const createClient = (options: {
  rpcUrl?: string;
  common?: Common;
  blockTag?: BlockTag | bigint;
}): MemoryClient => {
  const { rpcUrl, common, blockTag } = options;
  if (!rpcUrl) throw new Error("You need to provide a rpcUrl if you don't provide a client directly");

  return createMemoryClient({
    common,
    fork: {
      transport: http(rpcUrl),
      blockTag: blockTag ?? "latest",
    },
    miningConfig: { type: "manual" },
  });
};

export const getUnifiedParams = async <
  TAbi extends Abi | readonly unknown[] = Abi,
  TFunctionName extends ContractFunctionName<TAbi> = ContractFunctionName<TAbi>,
>(
  args: TraceStorageAccessOptions & TraceStorageAccessTxParams<TAbi, TFunctionName> & { config?: ExploreStorageConfig },
): Promise<TraceStorageAccessTxWithData & { client: MemoryClient }> => {
  const { client: _client, rpcUrl, common } = args;

  // Create the tevm client
  const client = _client ?? createClient({ rpcUrl, common });

  // Return early if the tx was already provided in calldata format
  if (args.from && args.data) return { client, from: args.from, to: args.to, data: args.data, value: args.value };

  // Encode calldata if the contract call was provided (abi, functionName, args)
  if (args.from && args.to && args.abi && args.functionName && args.args) {
    try {
      // @ts-expect-error complex union type not exactly similar
      const data = encodeFunctionData(args);
      return { client, from: args.from, to: args.to, data, value: args.value };
    } catch (err) {
      debug(`Failed to encode function data: ${err}`);
      throw err;
    }
  }

  // If we provide a value but no txHash, it's a simple transfer
  if (args.value && !args.txHash) return { client, from: args.from, to: args.to, value: args.value };

  // In this case, we need to replay the transaction
  if (!args.txHash)
    throw new Error("You need to provide a txHash if you don't provide the transaction data or contract call");

  // If we're replaying a transaction, extract the from, to, and data from the transaction
  try {
    const tx = await client.getTransaction({ hash: args.txHash });

    // Duplicate the client but fork before the transaction's block
    const clientBeforeTx = createMemoryClient({
      ...client,
      fork: {
        transport: client.transport,
        // Ideally we'd want to fork just right before that tx as some same-block txs can impact the state
        blockTag: tx.blockNumber > 0 ? tx.blockNumber - BigInt(1) : BigInt(0),
      },
    });

    return {
      client: clientBeforeTx,
      from: tx.from,
      to: tx.to ?? undefined,
      // TODO: remove when correctly formatted (tx in block mined here has data instead of input)
      // @ts-expect-error Property 'data' does not exist on type Transaction
      data: tx.input ? (tx.input as Hex) : (tx.data as Hex),
      value: tx.value,
    };
  } catch (err) {
    debug(`Failed to get transaction for replaying ${args.txHash}: ${err}`);
    throw err;
  }
};

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
