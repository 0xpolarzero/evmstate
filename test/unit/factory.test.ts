import { assert, describe, expect, it } from "vitest";

import { ACCOUNTS, CONTRACTS } from "@test/constants.js";
import { getClient } from "@test/utils.js";
import { traceState } from "@/index.js";

const { Factory, SimpleContract } = CONTRACTS;
const { caller } = ACCOUNTS;

/**
 * Contract creation tests
 *
 * This test suite verifies:
 *
 * 1. Tracing contract creation from a factory contract
 * 2. Capturing state changes in the factory contract (array updates)
 * 3. Capturing state changes in the created contracts (initialization)
 * 4. Tracing interactions with created contracts
 */
describe("Contract creation", () => {
  it("should trace contract creation and factory state changes", async () => {
    const client = getClient();

    // Deploy a contract from the factory
    expect(
      await traceState({
        ...Factory.write.createContract(123n),
        client,
        from: caller.toString(),
      }),
    ).toMatchSnapshot();
  });

  it("should trace interactions with a created contract", async () => {
    const client = getClient();

    // First create a contract
    const { data: contractAddress } = await client.tevmContract({
      ...Factory.write.createContract(123n),
      from: caller.toString(),
      addToBlockchain: true,
    });
    assert(contractAddress, "Contract address should be defined");

    // Trace a setter function call with the created contract
    expect(
      await traceState({
        ...SimpleContract.write.setValue(456n),
        client,
        from: caller.toString(),
        to: contractAddress,
      }),
    ).toMatchSnapshot();

    // Trace a getter function call with the created contract
    expect(
      await traceState({
        client,
        from: caller.toString(),
        to: contractAddress,
        abi: SimpleContract.abi,
        functionName: "getValue",
        args: [],
      }),
    ).toMatchSnapshot();
  });
});
