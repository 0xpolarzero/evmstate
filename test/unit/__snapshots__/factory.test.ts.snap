// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contract creation > should trace contract creation and factory state changes 1`] = `
{
  "0x0000000000000000000000000000000000000000": {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
  "0x0000000000000000000000000000000000000001": {
    "balance": {
      "current": 10000000000000000000n,
      "modified": true,
      "next": 9999999999998730585n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": true,
      "next": 1,
    },
    "storage": {},
  },
  "0x466163746f727900000000000000000000000000": {
    "balance": {
      "current": 0n,
      "modified": false,
    },
    "code": {
      "current": "0x608060405234801561000f575f5ffd5b5060043610610034575f3560e01c80639db8d7d514610038578063b32d4d4514610068575b5f5ffd5b610052600480360381019061004d919061022a565b610086565b60405161005f9190610294565b60405180910390f35b610070610169565b60405161007d9190610364565b60405180910390f35b5f5f826040516020016100999190610393565b60405160208183030381529060405290505f6040518061020001604052806101c781526020016104226101c79139826040516020016100d99291906103fe565b60405160208183030381529060405290505f8151602083015ff09050803b6100ff575f5ffd5b5f81908060018154018082558091505060019003905f5260205f20015f9091909190916101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550809350505050919050565b60605f8054806020026020016040519081016040528092919081815260200182805480156101e957602002820191905f5260205f20905b815f9054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190600101908083116101a0575b5050505050905090565b5f5ffd5b5f819050919050565b610209816101f7565b8114610213575f5ffd5b50565b5f8135905061022481610200565b92915050565b5f6020828403121561023f5761023e6101f3565b5b5f61024c84828501610216565b91505092915050565b5f73ffffffffffffffffffffffffffffffffffffffff82169050919050565b5f61027e82610255565b9050919050565b61028e81610274565b82525050565b5f6020820190506102a75f830184610285565b92915050565b5f81519050919050565b5f82825260208201905092915050565b5f819050602082019050919050565b6102df81610274565b82525050565b5f6102f083836102d6565b60208301905092915050565b5f602082019050919050565b5f610312826102ad565b61031c81856102b7565b9350610327836102c7565b805f5b8381101561035757815161033e88826102e5565b9750610349836102fc565b92505060018101905061032a565b5085935050505092915050565b5f6020820190508181035f83015261037c8184610308565b905092915050565b61038d816101f7565b82525050565b5f6020820190506103a65f830184610384565b92915050565b5f81519050919050565b5f81905092915050565b8281835e5f83830152505050565b5f6103d8826103ac565b6103e281856103b6565b93506103f28185602086016103c0565b80840191505092915050565b5f61040982856103ce565b915061041582846103ce565b9150819050939250505056fe6080604052348015600e575f5ffd5b506040516101c73803806101c78339818101604052810190602e9190606b565b805f81905550506091565b5f5ffd5b5f819050919050565b604d81603d565b81146056575f5ffd5b50565b5f815190506065816046565b92915050565b5f60208284031215607d57607c6039565b5b5f6088848285016059565b91505092915050565b6101298061009e5f395ff3fe6080604052348015600e575f5ffd5b50600436106030575f3560e01c8063209652551460345780635524107714604e575b5f5ffd5b603a6066565b60405160459190608d565b60405180910390f35b606460048036038101906060919060cd565b606e565b005b5f5f54905090565b805f8190555050565b5f819050919050565b6087816077565b82525050565b5f602082019050609e5f8301846080565b92915050565b5f5ffd5b60af816077565b811460b8575f5ffd5b50565b5f8135905060c78160a8565b92915050565b5f6020828403121560df5760de60a4565b5b5f60ea8482850160bb565b9150509291505056fea2646970667358221220587a82a5cd5d6ac106b4cd4612863e3e1d4be092d681218aa1abd6b9104a77bb64736f6c634300081c00330000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": true,
      "next": 1,
    },
    "storage": {
      "createdContracts": {
        "kind": "dynamic_array",
        "name": "createdContracts",
        "trace": [
          {
            "current": {
              "decoded": 0n,
              "hex": "0x00",
            },
            "fullExpression": "createdContracts._length",
            "modified": true,
            "next": {
              "decoded": 1n,
              "hex": "0x01",
            },
            "path": [
              {
                "kind": "array_length",
                "name": "_length",
              },
            ],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
          {
            "current": {
              "decoded": "0x0000000000000000000000000000000000000000",
              "hex": "0x00",
            },
            "fullExpression": "createdContracts[0]",
            "modified": true,
            "next": {
              "decoded": "0x89b0d1b5E6b53A1bA16cF7b6Ad34292943671EaE",
              "hex": "0x89b0d1b5e6b53a1ba16cf7b6ad34292943671eae",
            },
            "path": [
              {
                "index": 0n,
                "kind": "array_index",
              },
            ],
            "slots": [
              "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563",
            ],
          },
        ],
        "type": "address[]",
      },
    },
  },
  "0x89b0d1b5e6b53a1ba16cf7b6ad34292943671eae": {
    "storage": {
      "slot_0x0000000000000000000000000000000000000000000000000000000000000000": {
        "name": "slot_0x0000000000000000000000000000000000000000000000000000000000000000",
        "trace": [
          {
            "fullExpression": "slot_0x0000000000000000000000000000000000000000000000000000000000000000",
            "modified": false,
            "next": {
              "hex": "0x7b",
            },
            "note": "Could not label this slot access because no layout was found.",
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
      },
    },
  },
}
`;

exports[`Contract creation > should trace interactions with a created contract 1`] = `
{
  "0x0000000000000000000000000000000000000000": {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
  "0x0000000000000000000000000000000000000001": {
    "balance": {
      "current": 9999999999998730585n,
      "modified": true,
      "next": 9999999999998544175n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 1,
      "modified": true,
      "next": 2,
    },
    "storage": {},
  },
  "0x89b0d1b5e6b53a1ba16cf7b6ad34292943671eae": {
    "balance": {
      "current": 0n,
      "modified": false,
    },
    "code": {
      "current": "0x6080604052348015600e575f5ffd5b50600436106030575f3560e01c8063209652551460345780635524107714604e575b5f5ffd5b603a6066565b60405160459190608d565b60405180910390f35b606460048036038101906060919060cd565b606e565b005b5f5f54905090565b805f8190555050565b5f819050919050565b6087816077565b82525050565b5f602082019050609e5f8301846080565b92915050565b5f5ffd5b60af816077565b811460b8575f5ffd5b50565b5f8135905060c78160a8565b92915050565b5f6020828403121560df5760de60a4565b5b5f60ea8482850160bb565b9150509291505056fe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "modified": false,
    },
    "nonce": {
      "current": 1,
      "modified": false,
    },
    "storage": {
      "slot_0x0000000000000000000000000000000000000000000000000000000000000000": {
        "name": "slot_0x0000000000000000000000000000000000000000000000000000000000000000",
        "trace": [
          {
            "current": {
              "hex": "0x7b",
            },
            "fullExpression": "slot_0x0000000000000000000000000000000000000000000000000000000000000000",
            "modified": true,
            "next": {
              "hex": "0x01c8",
            },
            "note": "Could not label this slot access because no layout was found.",
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
      },
    },
  },
}
`;

exports[`Contract creation > should trace interactions with a created contract 2`] = `
{
  "0x0000000000000000000000000000000000000000": {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
  "0x0000000000000000000000000000000000000001": {
    "balance": {
      "current": 9999999999998730585n,
      "modified": true,
      "next": 9999999999998566274n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 1,
      "modified": true,
      "next": 2,
    },
    "storage": {},
  },
  "0x89b0d1b5e6b53a1ba16cf7b6ad34292943671eae": {
    "balance": {
      "current": 0n,
      "modified": false,
    },
    "code": {
      "current": "0x6080604052348015600e575f5ffd5b50600436106030575f3560e01c8063209652551460345780635524107714604e575b5f5ffd5b603a6066565b60405160459190608d565b60405180910390f35b606460048036038101906060919060cd565b606e565b005b5f5f54905090565b805f8190555050565b5f819050919050565b6087816077565b82525050565b5f602082019050609e5f8301846080565b92915050565b5f5ffd5b60af816077565b811460b8575f5ffd5b50565b5f8135905060c78160a8565b92915050565b5f6020828403121560df5760de60a4565b5b5f60ea8482850160bb565b9150509291505056fe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "modified": false,
    },
    "nonce": {
      "current": 1,
      "modified": false,
    },
    "storage": {
      "slot_0x0000000000000000000000000000000000000000000000000000000000000000": {
        "name": "slot_0x0000000000000000000000000000000000000000000000000000000000000000",
        "trace": [
          {
            "current": {
              "hex": "0x7b",
            },
            "fullExpression": "slot_0x0000000000000000000000000000000000000000000000000000000000000000",
            "modified": false,
            "note": "Could not label this slot access because no layout was found.",
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
      },
    },
  },
}
`;
